gc/include/omtalk/Barrier.h

```diff
@@ -11,34 +13,59 @@ namespace omtalk::gc {
 template <typename S, typename ObjectProxyT>
 void allocateBarrier(Context<S> &context, ObjectProxyT object) {
   // Newly allocated objects must be marked black and scanned by the GC.
-  if (context.writeBarrierEnabled()) {
-    mark<S>(context.getCollectorContext(), object);
-  }
+  mark<S>(context.getCollectorContext(), object);
 }
 
-template <typename S, typename ObjectProxyT, typename SlotProxyT>
-auto load(Context<S> &context, ObjectProxyT &object, SlotProxyT &slot) {
+template <typename S, typename SlotProxyT>
+auto load(Context<S> &context, SlotProxyT &slot) {
+
   // if the slot points to an evactuate region, copy the object into the current
   // allocation region.
-  auto result = slot.load();
-  if (context.loadBarrierEnabled()) {
-    auto region = Region::get(result);
-    if (region.isEvacuating()) {
-      // auto newObject =
-      OMTALK_ASSERT_UNREACHABLE();
+  Ref<void> address = slot.load();
+  auto *region = Region::get(address);
+  if (region->isEvacuating()) {
+    ForwardingMap &map = region->getForwardingMap();
+    auto &entry = map[address.get()];
+    if (entry.tryLock()) {
+      // Copy the object to the local allocation buffer.
+      GlobalCollectorContext<S> &gcContext = context.getCollectorContext();
+      auto &buffer = context.buffer();
+      auto *to = buffer.begin;
+      auto available = buffer.available();
+      auto result = copy(gcContext, ObjectProxy<S>(address), to, available);
+
+      if (!result) {
+        // get a new allocation buffer
+        auto copySize = getCopySize<S>(address);
+        context.refreshBuffer(copySize);
+        // Refreshing the buffer may block the thread until the end of the
+        // garbage collection cycle.
+        to = buffer.begin;
+        available = buffer.available();
+        result = copy(gcContext, address, to, available);
+      }
+
+      entry.set(to);
+      address = to;
+    } else {
+      // Object is being or has already been copied. entry.get() will wait for
+      // the object to be copied.
+      address = entry.get();
```

Comment:
If the thread doing the copying stalls out for an allocation buffer, this thread will never hit a gc sync point and the whole thing will deadlock.